Docs Rewrite (#87)
hirethunk/verbs: docs/testing.md

We enjoy improving Verbs by providing easy, readable testing affordances.

> ## `Verbs::Commit()`

Reviewer: ```suggestion
## `Verbs::commit()`
```

When testing verbs events, you'll need to call [commit](/docs/reference/events#content-committing) manually.

You may continue manually adding `Verbs::commit()` after each `Event::fire()` method; however, we've created `Verbs::commitImmediately` to issue a blanket commit on all events you fire in tests.

```php
beforeEach(function () {
    Verbs::commitImmediately();
});
```

You may also implement the `CommitsImmediately` interface directly on an Event.
(more about this in [`VerbsStatesInitialized`](testing#content-verbsstateinitialized))

> ### Assertions

Reviewer: These are only available if you call `Verbs::fake()` — we should document that.

The following Test `assert()` methods are available to thoroughly check your committing granularly.

```php
Verbs::assertNothingCommitted();
Verbs::assertCommitted(...);
Verbs::assertNotCommitted(...);
```

## State Factories

In tests, you may find yourself needing to fire and commit several events in order to bring your State to the point where it actually needs testing.

> The `ExampleState::factory()` method allows you to bypass manually building up the State, functioning similarly to `EloquentModel::factory()->state(['data'=>'data'])->create()`

Reviewer: ```suggestion
The `State::factory()` method allows you to bypass manually building up the State, functioning similarly to `Model::factory()`
```

This allows you to call:

```php
BankAccountState::factory()->create(
    data: ['balance' => 1337]
    id: $bank_account_id
);

// Or, using `id()` syntax:

BankAccountState::factory()
    ->id($bank_account_id)
    ->create(
        data: ['balance' => 1337]
    );
```

- If you accidentally pass an ID into both `id()` and `create()`, `create()` takes precedence.

Or, in the case of a [singleton state](/docs/reference/states#content-singleton-states):

```php
ChurnState::factory()->create(['churn' => 40]);
```

Next, we'll get into how these factories work, and continue after with some [Verbs factory methods](testing#content-factory-methods) you may already be familiar with from Eloquent factories.

### `VerbsStateInitialized`

Under the hood, these methods will fire (and immediately commit) a new `VerbsStateInitialized` event, which will fire onto the given state, identified by the id argument (if id is null, we assume it is a singleton) and return a copy of that state.

> This is primarily designed for booting up states for testing, but it may also be useful in migrating non event-sourced codebases to Verbs, when there is a need to initiate a state for legacy data.

Reviewer: I don't think we should suggest using this in prod. It's better to create your own "migrated from legacy" event in that case.

}
```

`VerbsStateInitialized` implements the `CommitsImmediately` interface detailed [above](testing#content-verbscommit), so if you change from this initial event makes sure to extend the interface on your replacement event.

### Factory Methods

Some methods accept Verbs [IDs](/docs/technical/ids), which, written longform, could be any of these types: `Bits|UuidInterface|AbstractUid|int|string`.

For brevity, this will be abbreviated in the following applicable methods as `Id`.

#### `count(int $count)`

Number of states to create. Returns a `StateCollection`.

```php
UserState::factory()->count(3)->create();
```

#### `id(Id $id)`

Set the state ID explicitly (cannot be used with `count`).

```php
UserState::factory()->id(123)->create();
```

#### `singleton()`

Mark that this is a singleton state (cannot be used with `count`).

```php
UserState::factory()->singleton()->create();
```

#### `state(callable|array $data)`

Default data (will be overridden by `create`).

```php
UserState::factory()->state([ /* state data */ ])->create();
```

The state function is mostly useful for [custom factories](#content-custom-factories).

> #### `create(Id|null $id = null))`

Reviewer: ```suggestion
#### `create(array $data, Id|null $id = null)`
```
